File: Native/FunctionReflection.cpp
#include "FunctionReflection.h"
#include "TypeReflection.h"
#include "VariableReflection.h"
#include "Attribute.h"
#include "GenericReflection.h"
#include "Modifier.h"
#include <stdexcept>

Native::FunctionReflection::FunctionReflection(void* native)
{
	m_native = (slang::FunctionReflection*)native;
}

slang::FunctionReflection* Native::FunctionReflection::getNative()
{
    return m_native;
}

char const* Native::FunctionReflection::getName()
{
    return m_native->getName();
}

Native::TypeReflection* Native::FunctionReflection::getReturnType()
{
    return new Native::TypeReflection(m_native->getReturnType());
}

unsigned int Native::FunctionReflection::getParameterCount()
{
    return m_native->getParameterCount();
}

Native::VariableReflection* Native::FunctionReflection::getParameterByIndex(unsigned int index)
{
    return new Native::VariableReflection(m_native->getParameterByIndex(index));
}

unsigned int Native::FunctionReflection::getUserAttributeCount()
{
    return m_native->getUserAttributeCount();
}
Native::Attribute* Native::FunctionReflection::getUserAttributeByIndex(unsigned int index)
{
    return new Native::Attribute(m_native->getUserAttributeByIndex(index));
}
Native::Attribute* Native::FunctionReflection::findAttributeByName(char const* name)
{
    return new Native::Attribute(m_native->findAttributeByName(SessionCLI::GetGlobalSession(), name));
}
Native::Attribute* Native::FunctionReflection::findUserAttributeByName(char const* name)
{
    return new Native::Attribute(m_native->findUserAttributeByName(SessionCLI::GetGlobalSession(), name));
}

Native::Modifier* Native::FunctionReflection::findModifier(Modifier::ID id)
{
    return new Native::Modifier(m_native->findModifier((slang::Modifier::ID)(int)id));
}

Native::GenericReflection* Native::FunctionReflection::getGenericContainer()
{
    return new Native::GenericReflection(m_native->getGenericContainer());
}

Native::FunctionReflection* Native::FunctionReflection::applySpecializations(GenericReflection* genRef)
{
    return new Native::FunctionReflection(m_native->applySpecializations((slang::GenericReflection*)genRef->getNative()));
}

Native::FunctionReflection* Native::FunctionReflection::specializeWithArgTypes(unsigned int argCount, TypeReflection* const* types)
{
    slang::TypeReflection** slangTypes = new slang::TypeReflection * [argCount];
    for (unsigned int i = 0; i < argCount; ++i) {
        slangTypes[i] = (slang::TypeReflection*)types[i]->getNative();
    }

	slang::FunctionReflection* specialized = m_native->specializeWithArgTypes(argCount, slangTypes);
	delete[] slangTypes; // Clean up the allocated array
    return new Native::FunctionReflection(specialized);
}

bool Native::FunctionReflection::isOverloaded()
{
    return m_native->isOverloaded();
}

unsigned int Native::FunctionReflection::getOverloadCount()
{
    return m_native->getOverloadCount();
}

Native::FunctionReflection* Native::FunctionReflection::getOverload(unsigned int index)
{
    return new Native::FunctionReflection(m_native->getOverload(index));
}

File: src/Slang.Sdk/Layer2-Binding/Reflection-API/FunctionReflection.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Slang.Sdk.Interop;
using static Slang.Sdk.Interop.StrongInterop;
using static Slang.Sdk.Interop.SlangNativeInterop;
using static Slang.Sdk.Interop.StringMarshaling;

namespace Slang.Sdk.Binding
{
    internal unsafe class FunctionReflection : Reflection, IEquatable<FunctionReflection>
    {
        internal override Reflection? Parent { get; }
        internal override FunctionReflectionHandle Handle { get; }
        internal override FunctionReflectionHandle NativeHandle => new(StrongInterop.FunctionReflection.GetNative(Handle, out var _));


        internal FunctionReflection(Reflection parent, FunctionReflectionHandle handle)
        {
            Parent = parent;
            Handle = handle;
        }

        internal string GetName()
        {
            string? error = null;
            // Test string?
            return Call(() => StrongInterop.FunctionReflection.GetName(Handle, out error), () => error);
        }

        internal TypeReflection GetReturnType()
        {
            string? error = null;
            return new TypeReflection(this, Call(() => StrongInterop.FunctionReflection.GetReturnType(Handle, out error), () => error));
        }

        internal uint GetParameterCount()
        {
            string? error = null;
            return Call(() => StrongInterop.FunctionReflection.GetParameterCount(Handle, out error), () => error);
        }

        internal VariableReflection GetParameterByIndex(uint index)
        {
            string? error = null;
            return new VariableReflection(this, Call(() => StrongInterop.FunctionReflection.GetParameterByIndex(Handle, index, out error), () => error));
        }

        internal ModifierReflection FindModifier(int modifierId)
        {
            string? error = null;
            return new ModifierReflection(this, Call(() => StrongInterop.FunctionReflection.FindModifier(Handle, modifierId, out error), () => error));
        }

        internal uint GetUserAttributeCount()
        {
            string? error = null;
            return Call(() => StrongInterop.FunctionReflection.GetUserAttributeCount(Handle, out error), () => error);
        }

        internal AttributeReflection GetUserAttributeByIndex(uint index)
        {
            string? error = null;
            return new AttributeReflection(this, Call(() => StrongInterop.FunctionReflection.GetUserAttributeByIndex(Handle, index, out error), () => error));
        }

        internal AttributeReflection FindAttributeByName(string name)
        {
            string? error = null;
            return new AttributeReflection(
                this,
                Call(() => StrongInterop.FunctionReflection.FindAttributeByName(Handle, name, out error), () => error));
        }

        internal GenericReflection GetGenericContainer()
        {
            string? error = null;
            return new GenericReflection(this, Call(() => StrongInterop.FunctionReflection.GetGenericContainer(Handle, out error), () => error));
        }

        internal FunctionReflection ApplySpecializations(GenericReflection genRef)
        {
            string? error = null;
            return new FunctionReflection(this, Call(() => StrongInterop.FunctionReflection.ApplySpecializations(Handle, genRef.Handle, out error), () => error));
        }

        internal FunctionReflection SpecializeWithArgTypes(uint typeCount, void** types)
        {
            string? error = null;
            return new FunctionReflection(this, Call(() => StrongInterop.FunctionReflection.SpecializeWithArgTypes(Handle, typeCount, types, out error), () => error));
        }

        internal bool IsOverloaded()
        {
            string? error = null;
            return Call(() => StrongInterop.FunctionReflection.IsOverloaded(Handle, out error), () => error);
        }

        internal uint GetOverloadCount()
        {
            string? error = null;
            return Call(() => StrongInterop.FunctionReflection.GetOverloadCount(Handle, out error), () => error);
        }

        internal FunctionReflection GetOverload(uint index)
        {
            string? error = null;
            return new FunctionReflection(this, Call(() => StrongInterop.FunctionReflection.GetOverload(Handle, index, out error), () => error));
        }

        public bool Equals(FunctionReflection? other)
        {
            return this == other;
        }
    }
}
